### PR TITLE
feat: Unit tests for dbt models

### LINKED ISSUE #102: Add accuracy tests against a known-good dataset
As the schema settles, we should add more testing around making sure the calculations performed in this repo are good / don't accidentally break with future updates. I think the following is a good start, but am open to better ideas:

- Create a fairly substantial dataset using the xapi-db-load tool's CSV output, maybe 100k rows
- Store those in a test folder here, in a folder like `test-seeds`
- As part of CI, rename that folder to `seeds` and load those files as seeds to the appropriate tables
- Write new dbt tests tagged with `ci` so we can explicitly check the outputs of views and MVs but turn it [off by default](https://docs.getdbt.com/reference/node-selection/yaml-selectors#default)
- In addition to the normal `dbt test` run, we should also run a `dbt test --select tag:ci` (I'm not 100% sure this will work with default off) and ensure the additional tests all come out correctly
- It would be nice to reload the seeds and re-run the tests to check that all of our deduplication is working as well 

Where possible it would be great if the tests can be assumed to work outside of the known-good dataset, in which case we should exclude marking them with the ci tag.

### PR DESCRIPTION
**Pull request by saraburns1**
Add unit tests for all models using a known good data set

Closes #102 

notes:
- default selector cannot exclude tests or else it excludes any models that have a test associated with them
- default schema macro needed to be updated to use custom schema for seeds in order for our sources to work

### CODE REVIEW

File: macros/get_problem_id.sql [hunk 1]
```diff
@@ -3,6 +3,6 @@
 -- or is followed by '/answer' or '/hint'
 {% macro get_problem_id(object_id) %}
     regexpExtract(
-        object_id, 'xblock/([\w\d-\+:@]*@problem\+block@[\w\d][^_]*)(_\d_\d)?', 1
+        object_id, 'xblock/([\w\d-\+:@]*@problem\+block@[\w\d][^_\/]*)(_\d_\d)?', 1
```
Comment (saraburns1) [inline macros/get_problem_id.sql:6]: This regexp was not removing '/answer' from the end of the string which caused data to be excluded in subsequent queries

File: models/video/unit_tests.yaml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      - input: ref('int_videos_per_subsection')
+        format: sql
+        rows: |
+          select * from int_videos_per_subsection
+    expect:
+      format: sql
+      rows: |
+        select * from section_video_engagement_expected
+
+  # - name: test_subsection_video_engagement
```
Comment (saraburns1) [inline models/video/unit_tests.yaml:90]: these are having an issue I can't figure out right now but I wanted to get the rest of this code up for review

File: .DS_Store
Comment (Ian2012) [inline .DS_Store:1]: please remove this file

File: models/base/sources.yml [hunk 3]
```diff
@@ -12,3 +12,11 @@ sources:
...
           - name: emission_time
           - name: event
+  - name: reporting
+    database: "reporting"
+    tables:
+      - name: xapi_events_all
+        columns:
+          - name: event_id
+          - name: emission_time
+          - name: event
```
Comment (Ian2012) [inline models/base/sources.yml:22]: why is this model added here?

I think you need to add the environment variables for the databases to the CI workflow:
```yaml
ASPECTS_EVENT_SINK_DATABASE: event_sink
ASPECTS_XAPI_DATABASE: xapi
DBT_PROFILE_TARGET_DATABASE: reporting 
```

File: .DS_Store
Comment (saraburns1) [reply .DS_Store:1]: I have no idea where this is coming from, I don't see it locally so I think we just need to add it to the repo's gitignore file

File: macros/source_for_test.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{% macro source_for_test(schema, table) %}
+
+    {% if target.name != "prod" %} {{ source(target.schema, table) }}
+    {% else %} {{ source(schema, table) }}
+    {% endif %}
+
+{% endmacro %}
```
Comment (Ian2012) [inline macros/source_for_test.sql:7]: why is this needed?

In case this is caused by the discrepancy on the settings for databases, I think we can remove it

Comment (saraburns1) [reply macros/source_for_test.sql:7]: Correct. The models use sources like `from {{ source("xapi", "xapi_events_all") }}` but dbt creates the seed tables in the database specified in the profiles.yml file which is different depending on if you are running locally or in CI

File: models/courses/sources.yml [hunk 5]
```diff
@@ -32,3 +32,67 @@ sources:
...
+          - name: course_start
+          - name: course_end
+          - name: enrollment_start
+          - name: enrollment_end
+          - name: self_paced
+          - name: course_data_json
+          - name: created
+          - name: modified
+          - name: dump_id
+          - name: time_last_dumped
```
[the remark this answers is not in the document]
Comment (saraburns1) [reply models/courses/sources.yml:98]: Oooh that's a good point, hadn't thought of that...

File: macros/format_float_value.sql [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{% macro format_float_value(value) %}
+
+    {% if target.name != "prod" %} CAST({{ value }} as Float32)
+    {% else %} {{ value }}
+    {% endif %}
```
Comment (Ian2012) [inline macros/format_float_value.sql:5]: why is this conditional?

Comment (saraburns1) [reply macros/format_float_value.sql:5]: I didn't want to change the actual query output, but dbt does something weird with floats and had trouble comparing the data in the tests

**Review by Ian2012 — COMMENTED** [review]
Just a couple of comments with the new sources and macros

File: macros/format_float_value.sql (hunk 6 above)
Comment (bmtcril) [reply macros/format_float_value.sql:5]: Is it the values in the seed tables that are causing problems? I wonder if there's something we can do to get them inserted correctly. 🤔

Comment (saraburns1) [reply macros/format_float_value.sql:5]: No, it's something within the dbt code that does the comparison. No matter what I changed the seeded values to, it always thought they were wrong. It seems like dbt is casting values as a mixture of decimal and float as it runs which changes the end decimals in float64. The first 6 or so digits are the same, but the rest are different from anything I could reproduce.

Comment (bmtcril) [reply macros/format_float_value.sql:5]: What I'm seeing is that:

- Initial grading event in `xapi_events_all.csv` and the table as a value of `0.9523809523809523`. That value seems to carry through the whole stream and is what's expected at the end. 
- `grading_events_expected.csv` has a value of `0.95238096` for that event
- `fact_learner_course_grade_expected.csv` / `fact_student_status_expected.csv` have a value of `0.9523809552192688`

I've been testing with removing the format_float_value macro on `fact_student_status` and if I update the expected value to match what's in `xapi_events_all/.csv` the tests pass. My best guess is that the values are getting lost as part of the process of exporting the expected csv values. What are you doing to generate those? 

It would probably be good to document that as well, for when we need to add more tests later.

Comment (saraburns1) [reply macros/format_float_value.sql:5]: Here's what I did:

I have xapi_all_events_parsed in my local clickhouse.
Event_id b47f3b76-21a8-45ec-b219-f11ef05739df shows score of 0.9523809523809523

Run grading_events query against that table
`
select
    event_id,
    CAST(emission_time, 'DateTime') as emission_time,
    actor_id,
    object_id,
    course_key,
    org,
    verb_id,
    JSONExtractFloat(event, 'result', 'score', 'scaled')
    as scaled_score
from xapi_events_all_parsed
where
    verb_id in (
        'http://id.tincanapi.com/verb/earned',
        'https://w3id.org/xapi/acrossx/verbs/evaluated'
    )
    or (
        verb_id in (
            'http://adlnet.gov/expapi/verbs/passed',
            'http://adlnet.gov/expapi/verbs/failed'
        )
        and JSON_VALUE(event::String, '$.object.definition.type')
        = 'http://adlnet.gov/expapi/activities/course'
    )
`
Export table from Clickhouse
Open in VSCode
Save as expected csv
Still has correct value in csv
image

Removed float macro from grading_events
dbt seed full refresh
dbt run full refresh
dbt test 

Test fails with:
image

Comment (saraburns1) [reply macros/format_float_value.sql:5]: I also tried changing the unit test query to
`select event_id, emission_time, actor_id, object_id, course_key, org, verb_id, cast(scaled_score as Float64) scaled_score from grading_events_expected`
but it made no difference, the incorrect value was still used for the test

Comment (bmtcril) [reply macros/format_float_value.sql:5]: Which test is it that's failing? 

I had a second change left over from earlier attempts that seems necessary to fix `test_fact_student_status.sql`, and that was a yaml to force the course_grade to a Float64:


unit-test-seeds/grading/grading_seeds.yaml:

```
version: 2

seeds:
  - name: fact_student_status_expected
    config:
      column_types:
        course_grade: Float64
```

Comment (bmtcril) [reply macros/format_float_value.sql:5]: I think we would need to re-export anything with wrong values in the expected files, then add the yaml to force any that are somehow being implicitly created as Float32. But I thiiiiiiink that's all we need to get around the macro?

Comment (saraburns1) [reply macros/format_float_value.sql:5]: The grading_events test is failing in what I sent. Adding that yaml for grading_events still doesn't work. I agree that changing the values on the other tests expected data would work, but they all stem from grading_events which fails without the float macro

Comment (saraburns1) [reply macros/format_float_value.sql:5]: Ahhhh I got it. You're right, we need the column_type specification. I was using course_grade and not scaled_score so it wasn't working for grading_events.

I'll update everything and remove the macro. Thanks for the help!!

**Review by bmtcril — APPROVED** [review]
LGTM, thanks for all of the work on this!
